### PR TITLE
Homework/04

### PR DESCRIPTION
**Pull request by ameyuuno**

### CODE REVIEW

File: src/main/scala/basics/Collections.scala [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  object FromCollection {
+    def min(list: List[Int]): Option[Int] = if (list.isEmpty) None else Some(list.reduce(math.min))
+
+    def scanLeft[T](zero: T)(list: List[T])(f: (T, T) => T): List[T] = {
+      @tailrec
+      def _scanLeft(values: List[T], accumulators: List[T]): List[T] = values match {
+        case Nil          => accumulators
+        case head :: tail => _scanLeft(tail, accumulators :+ f(accumulators.last, head))
```
Comment (nikalaikina) [inline src/main/scala/basics/Collections.scala:82]: I think it is better to add to the beginning of the list and then reverse it

File: src/main/scala/basics/Collections.scala [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import scala.annotation.tailrec
+
+object Collections {
+
+  object LeetCode1480 {
+    def runningSum(nums: Array[Int]): Array[Int] = nums.scanLeft(0)(_ + _).tail
+  }
+
+  object LeetCode1470 {
+    def shuffle(nums: Array[Int], n: Int): Array[Int] = (nums.take(n) zip nums.drop(n)).flatMap(pair => List(pair._1, pair._2))
```
Comment (nikalaikina) [inline src/main/scala/basics/Collections.scala:12]: Why you all love these _1 and _2 so much T_T

File: src/main/scala/basics/DataStructures.scala [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    .map { case (vegetable, amount) => (vegetable, vegetableWeights(vegetable) * amount) }
+
+  def allSubsetsOfSizeN[A](set: Set[A], n: Int): Set[Set[A]] =
+    n match {
+      case 0 => Set()
+      case 1 => set.map(Set(_))
+      case n =>
+        for {
+          element <- set
+          subsetExcludingElement <- allSubsetsOfSizeN(set.excl(element), n - 1)
```
Comment (nikalaikina) [inline src/main/scala/basics/DataStructures.scala:40]: `set - element` looks good to me

File: src/main/scala/basics/Collections.scala [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  object LeetCode1637 {
+    def maxWidthOfVerticalArea(points: Array[Array[Int]]): Int = points.map(_.head).sorted.sliding(2).map(pair => pair(1) - pair(0)).max
+  }
+
+  object LeetCode1614 {
+    def maxDepth(s: String): Int = {
+      @tailrec
+      def countMaxNestedParenthesesDepth(string: String, parenthesesDepth: Int = 0, maxParenthesesDepth: Int = 0): Int = string.headOption match {
```
Comment (nikalaikina) [inline src/main/scala/basics/Collections.scala:30]: I think separating finding depth's from finding max would make code simpler

**Review by nikalaikina — APPROVED** [review]

File: src/main/scala/basics/Collections.scala (hunk 2 above)
Comment (ameyuuno) [reply src/main/scala/basics/Collections.scala:12]: I do not like it >~< 
Mb replace it with { case (x, y) => … }? Is it better?

**Pull request merged**